st4/__init__.py

```diff
@@ -66,6 +72,33 @@ def m_metals_status(self, params: Any) -> None:
         else:
             session.erase_window_status_async(key)
 
+    def m_metals_publishDecorations(self, decorationsParams: Any) -> None:
+        if not isinstance(decorationsParams, dict):
+            return
+
+        session = self.weaksession()
+        if not session:
+            return
+
+        uri = decorationsParams.get('uri')
+        if not uri:
+            return
+
+        session_buffer = session.get_session_buffer_for_uri_async(uri)
+        if not session_buffer:
+            print("no session builder")
+            return
+
+        for sv in session_buffer.session_views:
+            phantom_set = self._phantom_sets.get(sv.view.id())
+
+            if not phantom_set:
+                phantom_set = sublime.PhantomSet(sv.view, self.phantom_key)
+                self._phantom_sets[sv.view.id()] = phantom_set
+
+            phantom_set.update(decorations_to_phantom(decorationsParams.get('options', []), sv.view))
```

Comment:
With respect to the lifetime management of the phantom sets: I've been thinking about this for a while now and while this may sound hacky it's okay to do it here because this plugin is a leaf in the tree of abstractions anyway (nobody is going to inherit/expand upon this code further). You can stash a `sublime.PhantomSet` inside of a `SessionBuffer` by doing something like this:

```python
phantom_set = getattr(sv, "_lsp_metals_decorations")
if not phantom_set:
    phantom_set = sublime.PhantomSet(sv.view, self.phantom_key)
    setattr(sv, "_lsp_metals_decorations", phantom_set)
phantom_set.update(...)
```

This will bound the lifetime of the phantom set to the lifetime of the session view, which is handled correctly by LSP.

Reply:
That thanks for the tip, I ended up using that